File: cp-library-rs/src/data_structure/union_find.rs
//! UnionFind木

use std::{collections::HashMap, fmt::Debug, mem};

use crate::utils::consts::NEG1;

/// UnionFind木
pub struct UnionFind {
    /// 要素数
    n: usize,
    /// 親の番号を格納する配列
    parent: Vec<usize>,
    /// 連結成分の個数
    count: usize,
}

impl UnionFind {
    /// UnionFindを新規作成
    pub fn new(N: usize) -> Self {
        UnionFind {
            n: N,
            parent: vec![NEG1; N],
            count: N,
        }
    }

    /// 根を求める
    pub fn root(&mut self, mut x: usize) -> usize {
        // 根を探索
        let mut root = x;
        while self.parent[root] < self.n {
            root = self.parent[root];
        }
        // 経路圧縮
        while self.parent[x] < self.n {
            x = mem::replace(&mut self.parent[x], root);
        }
        root
    }

    /// 同一の集合に所属するか判定
    pub fn is_same(&mut self, x: usize, y: usize) -> bool {
        self.root(x) == self.root(y)
    }

    /// 集合`x,y`を併合する．
    ///
    /// **戻り値**
    /// - すでに併合済みだった場合`None`，そうでない場合親となった要素の番号を返す
    pub fn unite(&mut self, x: usize, y: usize) -> Option<usize> {
        let mut parent = self.root(x);
        let mut child = self.root(y);

        if parent == child {
            return None;
        }

        // 要素数が大きい方を親にすることで、高さを均等に保つ
        if self.parent[parent] > self.parent[child] {
            (parent, child) = (child, parent);
        }

        self.parent[parent] = self.parent[parent].wrapping_add(self.parent[child]);
        self.parent[child] = parent;
        self.count -= 1;

        Some(parent)
    }

    /// 連結成分の大きさを求める
    pub fn get_size(&mut self, x: usize) -> usize {
        let root = self.root(x);
        self.parent[root].wrapping_neg()
    }

    /// 連結成分の数を返す
    pub fn group_count(&self) -> usize {
        self.count
    }

    /// {代表元: 集合} のマップを返す
    ///
    /// - 計算量: $`O(N)`$
    pub fn enum_groups(&mut self) -> HashMap<usize, Vec<usize>> {
        (0..self.n).fold(HashMap::default(), |mut map, i| {
            let root = self.root(i);
            map.entry(root).or_insert_with(Vec::new).push(i);
            map
        })
    }
}

impl Debug for UnionFind {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let mut uf = UnionFind {
            n: self.n,
            parent: self.parent.clone(),
            count: self.count,
        };
        let groups = uf.enum_groups();

        f.debug_map().entries(groups).finish()
    }
}

File: cp-library-py/cp_library_py/union_find.py
class UnionFind:
    def __init__(self, n: int) -> None:
        self._n = n

        # self._data[i] :=
        #     -x → i 番目の要素が根であり、グループに属する頂点数が x 個
        #      x → 要素 i の親が x
        self._data = [-1] * n

        # グループの個数
        self._group_count = n

    def find(self, u: int) -> int:
        """要素 u が属するグループの根を返す

        Args:
            u (int): 要素の番号

        Returns:
            int: 要素 u が属するグループの根
        """
        root = u

        while self._data[root] >= 0:
            root = self._data[root]

        while u != root:
            self._data[u], u = root, self._data[u]

        return root

    def is_same(self, u: int, v: int) -> bool:
        """要素 u と要素 v が同じグループに属するかを返す

        Args:
            u (int): 要素の番号
            v (int): 要素の番号

        Returns:
            bool: 要素 u と要素 v が同じグループに属するか
        """
        return self.find(u) == self.find(v)

    def unite(self, u: int, v: int) -> bool:
        """要素 u が属するグループと要素 v が属するグループを併合する

        Args:
            u (int): 要素の番号
            v (int): 要素の番号

        Returns:
            bool: 併合が行われたか
        """
        u = self.find(u)
        v = self.find(v)

        if u == v:
            return False

        if self._data[u] > self._data[v]:
            u, v = v, u

        self._data[u] += self._data[v]
        self._data[v] = u

        self._group_count -= 1

        return True

    def size(self, u: int) -> int:
        """要素 u が属するグループの頂点数を返す

        Args:
            u (int): 要素の番号

        Returns:
            int: 要素 u が属するグループの頂点数
        """
        return -self._data[self.find(u)]

    def group_count(self) -> int:
        """グループの個数を返す

        Returns:
            int: グループの個数
        """
        return self._group_count
